easybuild/tools/modules.py

```diff
@@ -543,9 +546,25 @@ def mod_exists_via_show(mod_name):
 
             :param mod_name: module name
             """
-            mod_exists_regex = mod_exists_regex_template % re.escape(mod_name)
-            txt = self.show(mod_name)
-            return bool(re.search(mod_exists_regex, txt, re.M))
+            stderr = self.show(mod_name)
+            res = False
+            # Parse the output:
+            # - Skip whitespace
+            # - Any error -> Module does not exist
+            # - Check first non-whitespace line for something that looks like an absolute path terminated by a colon
+            mod_exists_regex = r'\s*/.+:\s*'
+            for line in stderr.split('\n'):
+                if OUTPUT_MATCHES['whitespace'].search(line):
+                    continue
+                if OUTPUT_MATCHES['error'].search(line):
+                    break
+                if re.match(mod_exists_regex, line):
+                    res = True
+                break
```

Comment:
If there's no match, you'll just continue processing, and may hit a match later, which disagrees with what the comments above claim (first non-whitespace line).

Maybe use:

```python
            for line in stderr.split('\n'):
                if OUTPUT_MATCHES['whitespace'].search(line):
                    continue
                elif OUTPUT_MATCHES['error'].search(line):
                    break
                else:
                    res = bool(re.match(mod_exists_regex, line))
                    break
            return res

Reply:
Actually, I think skipping over non-matching values is correct, see the example in https://github.com/easybuilders/easybuild/issues/368#issuecomment-324910159: the `------` line must be part of `stderr` (since the stuff sent to `stdout` is supposed to be Python syntax).

So, you'll need to make sure you process all lines, not just stop at the first non-whitespace/error line.

```python
for line in stderr.split('\n'):
    if OUTPUT_MATCHES['whitespace'].search(line):
        continue
    elif OUTPUT_MATCHES['error'].search(line):
        break

    res = bool(re.match(mod_exists_regex, line))
    # if we found a match, we're done;
    # if not, we keep processing the "show" output
    if res:
        break

    return res
```

Wit this change, you should be able to reinstate the test you removed in https://github.com/easybuilders/easybuild-framework/pull/3216/commits/1454dd34da9bfc252f8c77337fcd3efc1892b1f7 ?

Reply:
> If there's no match, you'll just continue processing, and may hit a match later, which disagrees with what the comments above claim

No, you missed the break which is triggered for the first non-empty line. And only the first non-empty line must be checked, so your proposed code would be wrong (to permissive)

> Actually, I think skipping over non-matching values is correct, see the example in easybuilders/easybuild#368 (comment):

No. The python code goes to stdout. Previously (to this PR) the `show` method return `stdout + stderr` which explains the bug in #368.   
The `module` shell function works by `eval`ing stdout and printing stderr to stdout. This is what I'm using here

Reply:
The `------------` line, which is what your implementation was tripping over, must have been part of the `stderr` output, so that test case remains relevant now you're only grabbing the `stderr` output from `module show`?

So, you can't just stop processing the `show` stderr output lines after the first non-whitespace line that's not an error, or you'll base potentially base the result on an output line like `-------------` (which is why the test you removed was failing).

Reply:
Still no. The `OUTPUT_MATCHES['whitespace']` regex was designed to remove those `-------` lines (great idea BTW! wondering why this isn't used more)

The test stuffed in a mix of stdout+stderr into the cache which can never occur (anymore). The test then (rightly) failed due to ` "import os"` being the first non-empty line

Reply:
OK, got it, I forgot about the `whitespace` regex also handling `----` lines...